homeassistant/components/broadlink/__init__.py

```diff
@@ -1,125 +1,72 @@
-"""The broadlink component."""
-import asyncio
-from base64 import b64decode, b64encode
-from binascii import unhexlify
+"""The Broadlink integration."""
+from collections import namedtuple
 import logging
-import re
 
-from broadlink.exceptions import BroadlinkException, ReadError, StorageError
 import voluptuous as vol
 
-from homeassistant.const import CONF_HOST
-import homeassistant.helpers.config_validation as cv
-from homeassistant.util.dt import utcnow
-
-from .const import CONF_PACKET, DOMAIN, LEARNING_TIMEOUT, SERVICE_LEARN, SERVICE_SEND
-
-_LOGGER = logging.getLogger(__name__)
+from homeassistant import config_entries
+from homeassistant.const import (
+    CONF_DEVICES,
+    CONF_HOST,
+    CONF_MAC,
+    CONF_TIMEOUT,
+    CONF_TYPE,
+)
+from homeassistant.helpers import config_validation as cv
 
-DEFAULT_RETRY = 3
+from .const import DEFAULT_TIMEOUT, DOMAIN
+from .device import BroadlinkDevice
+from .helpers import mac_address
 
+LOGGER = logging.getLogger(__name__)
 
-def data_packet(value):
-    """Decode a data packet given for broadlink."""
-    value = cv.string(value)
-    extra = len(value) % 4
-    if extra > 0:
-        value = value + ("=" * (4 - extra))
-    return b64decode(value)
+DEVICE_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_HOST): cv.string,
+        vol.Inclusive(CONF_MAC, "manual_config"): mac_address,
+        vol.Inclusive(CONF_TYPE, "manual_config"): cv.positive_int,
+        vol.Optional(CONF_TIMEOUT, default=DEFAULT_TIMEOUT): cv.positive_int,
+    }
+)
 
+DOMAIN_SCHEMA = vol.Schema(
+    {vol.Optional(CONF_DEVICES): vol.All(cv.ensure_list, [DEVICE_SCHEMA])}
+)
 
-def hostname(value):
-    """Validate a hostname."""
-    host = str(value)
-    if len(host) > 253:
-        raise ValueError
-    if host[-1] == ".":
-        host = host[:-1]
-    allowed = re.compile(r"(?![_-])[a-z\d_-]{1,63}(?<![_-])$", flags=re.IGNORECASE)
-    if not all(allowed.match(elem) for elem in host.split(".")):
-        raise ValueError
-    return host
+CONFIG_SCHEMA = vol.Schema({DOMAIN: DOMAIN_SCHEMA}, extra=vol.ALLOW_EXTRA)
```

Comment:
Done. I was using the import as a fallback mechanism for discovery, in case the device does not say hello back and the user wants to set up the device manually. I'm not sure if anyone would need it.

Should I add a manual configuration step to the configuration flow? We could also wait to see if anyone really needs this option, so we avoid unnecessary code.

Reply:
Could we release a beta for mass testing before we finally merge on master? I'm sure the code is working, just not sure if all devices will respond "hello" correctly. If not, we should have a fallback mechanism.

Reply:
Done.